File: golang/2483_best_closing_time/best_closing_time.go
// Package main ...
package main

import (
	"testing"

	"github.com/stretchr/testify/assert"
)

func bestClosingTime(customers string) int {
	n := len(customers)
	suf := 0
	pre := 0
	minCost := 0
	res := 0

	for i := 0; i <= n; i++ {
		if minCost > suf+pre {
			minCost = suf + pre
			res = i
		}
		if i < n && customers[i] == 'N' {
			pre++
		} else if i < n {
			suf--
		}
	}
	return res
}

func main() {
	tests := []struct {
		customers string
		ans       int
	}{
		{"YYNY", 2},
		{"NNNNN", 0},
		{"YYYY", 4},
	}

	for index, test := range tests {
		assert.Equal(&testing.T{}, test.ans, bestClosingTime(test.customers), index)
	}
}

File: cpp/2483_best_closing_time/best_closing_time.cpp
#include <cassert>
#include <string>
#include <tuple>
#include <vector>

using namespace std;

class Solution {
 public:
  int bestClosingTime(string customers) {
    int n = customers.size();
    int suf = 0;
    int pre = 0;
    int min_cost = 0, res = 0;
    for (int i = 0; i <= n; i++) {
      if (min_cost > suf + pre) {
        min_cost = suf + pre;
        res = i;
      }
      if (i < n && customers[i] == 'N') {
        pre++;
      } else {
        suf--;
      }
    }
    return res;
  }
};

int main() {
  vector<tuple<string, int>> tests{
      {"YYNY", 2},
      {"NNNNN", 0},
      {"YYYY", 4},
  };

  for (auto& [customers, ans] : tests) {
    assert(Solution().bestClosingTime(customers) == ans);
  }
}